// @daraja/client/dist/client/daraja.client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const date_fns_1 = require("date-fns");
const errors_1 = require("./errors");
const schemas_1 = require("./schemas");
const utils_1 = require("./utils");
/**
 * This class contains methods that form and send requests to Daraja API
 */
class DarajaClient {
    /**
     * @param config.businessShortcode - the organizations shortcode (Paybill or Buygoods - A 5 to 6 digit account number) used to identify an organization and receive the transaction.
     * @param config.consumerKey - used as the username to authenticate requests for generating access tokens
     * @param config.consumerSecret - used as the username to authenticate requests for generating access tokens
     * @param config.environment - the environment to send requests to (defaults to 'sandbox')
     * @param config.scopes - an array representing the operations permitted for this instance of the client (defaults to an empty array)
     * @param config.lnmPasskey - a string issued by Safaricom used to generate the password required for Lipa na Mpesa Online requests (must be provided if the 'lipa-na-mpesa-online' scope is included)
     * @throws {@link DarajaClientError}
     * @example
     * ```typescript
     * const client = new DarajaClient({
     *   businessShortcode: '123456',
     *   consumerKey: 'consumerKey',
     *   consumerSecret: 'consumerSecret',
     *   lnmPasskey: 'lnmPasskey',
     *   scopes: ['lipa-na-mpesa-online'],
     * });
     * ```
     */
    constructor(config) {
        const { businessShortcode, consumerKey, consumerSecret, environment, scopes, lnmPasskey, } = (0, utils_1.validateObject)(config, schemas_1.darajaClientConfigSchema);
        this.validateScopes(scopes, lnmPasskey);
        const baseURL = `https://${environment === 'production' ? 'api' : 'sandbox'}.safaricom.co.ke`;
        this.businessShortcode = businessShortcode;
        this.authHttp = axios_1.default.create({
            baseURL: `${baseURL}/oauth`,
            params: { grant_type: 'client_credentials' },
            auth: { username: consumerKey, password: consumerSecret },
        });
        this.mpesaHttp = axios_1.default.create({ baseURL: `${baseURL}/mpesa` });
    }
    businessShortcode;
    accessTokenExpiry = new Date();
    lnmPasskey = undefined;
    authHttp;
    mpesaHttp;
    validateScopes = (scopes, lnmPasskey) => {
        if (scopes.includes('lipa-na-mpesa-online')) {
            if (lnmPasskey != null)
                this.lnmPasskey = lnmPasskey;
            else
                throw new errors_1.DarajaClientError(`'lnmPasskey' must be provided if scope includes 'lipa na mpesa-online'`);
        }
    };
    hasTokenExpired = () => (0, date_fns_1.isAfter)(new Date(), (0, date_fns_1.sub)(this.accessTokenExpiry, { minutes: 1 }));
    refreshAccessToken = async () => {
        if (this.hasTokenExpired()) {
            try {
                const { data } = await this.authHttp.get('/v1/generate');
                this.mpesaHttp.interceptors.request.use((0, utils_1.authInterceptor)(data.access_token));
                this.accessTokenExpiry = (0, date_fns_1.add)(new Date(), {
                    seconds: parseInt(data.expires_in),
                });
            }
            catch (error) {
                throw new errors_1.DarajaAuthenticationError(error.response?.statusText);
            }
        }
    };
    /**
     *
     * @internal
     */
    makeRequest = async (url, request) => {
        try {
            await this.refreshAccessToken();
            const { data } = await this.mpesaHttp.post(url, request);
            return data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error) && error.response != null)
                return error.response.data;
            throw error;
        }
    };
    /**
     *
     * @internal
     */
    generateTimestampAndLNMPassword = () => {
        if (this.lnmPasskey != null) {
            const timestamp = (0, date_fns_1.format)(new Date(), 'yyyyMMddHHmmss');
            const password = Buffer.from(`${this.businessShortcode}${this.lnmPasskey}${timestamp}`).toString('base64');
            return { timestamp, password };
        }
        throw new errors_1.DarajaClientError('Lipa na Mpesa Online Passkey is missing');
    };
    /**
     * @param options.accountReference - This is an alpha-numeric parameter that is defined by your system as an Identifier of the transaction for CustomerPayBillOnline transaction type. Along with the business name, this value is also displayed to the customer in the STK Pin Prompt message. Maximum of 12 characters.
     * @param options.amount - This is the amount transacted, normaly a numeric value. It is the money that customer pays to the Shorcode. Only whole numbers are supported.
     * @param options.callbackURL - This is a valid secure URL that is used to receive notifications from M-Pesa API. It is the endpoint to which the results will be sent by M-Pesa API.
     * @param options.partyA - This is the phone number sending money. The parameter expected is a Valid Safaricom Mobile Number that is M-Pesa registered in the format 2547XXXXXXXX
     * @param options.transactionDescription - This is any additional information/comment that can be sent along with the request from your system. Maximum of 13 Characters.
     * @param options.transactionType - This is the transaction type that is used to identify the transaction when sending the request to M-Pesa.
     * @param options.partyB - The organization receiving the funds. The parameter expected is a 5 to 7 digit (defaults to businessShortcode passed in the {@link DarajaClient} constructor)
     * @param options.phoneNumber - This is the Mobile Number to receive the STK Pin Prompt (defaults to partyA)
     * @throws {@link DarajaClientError} | {@link DarajaAuthenticationError}
     * @example
     * ```typescript
     * const transact = async () => {
     *   try {
     *     const response = await client.lipaNaMpesaOnlinePayment({
     *       accountReference: 'CompanyXLTD',
     *       amount: 1,
     *       callbackURL: 'http://mydomain.com/lnmcallback',
     *       partyA: '254701234567,
     *       transactionDescription: 'Payment of X',
     *       transactionType: 'CustomerPayBillOnline',
     *     });
     *     // do something with the response
     *   } catch (error) {
     *     // handle the error
     *   }
     * };
     *
     * transact()
     * ```
     */
    lipaNaMpesaOnlinePayment = async (options) => {
        const { timestamp, password } = this.generateTimestampAndLNMPassword();
        const { accountReference, amount, callbackURL, partyA, transactionDescription, transactionType, partyB, phoneNumber, } = (0, utils_1.validateObject)(options, schemas_1.lipaNaMpesaOnlinePaymentOptionsSchema);
        const data = await this.makeRequest('/stkpush/v1/processrequest', {
            AccountReference: accountReference,
            Amount: amount,
            BusinessShortCode: this.businessShortcode,
            CallBackURL: callbackURL,
            PartyA: partyA,
            PartyB: partyB ?? this.businessShortcode,
            Password: password,
            PhoneNumber: phoneNumber ?? partyA,
            Timestamp: timestamp,
            TransactionDesc: transactionDescription,
            TransactionType: transactionType,
        });
        return data;
    };
}
exports.default = DarajaClient;
